File: src/main/java/hu/gerviba/webschop/service/OrderService.java
package hu.gerviba.webschop.service;

import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

import javax.servlet.http.HttpServletRequest;
import javax.transaction.Transactional;

import hu.gerviba.webschop.dao.*;
import hu.gerviba.webschop.model.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import hu.gerviba.webschop.web.ControllerUtil;
import hu.gerviba.webschop.web.comonent.CustomComponentType;

import static java.util.stream.Collectors.groupingBy;

@Service
@Transactional
public class OrderService {

    @Autowired
    OrderRepository repo;
    
    @Autowired
    OpeningRepository openingRepo;
    
    @Autowired
    ControllerUtil util;
    
    @Autowired
    OpeningService openings;

    @Autowired
    TimeWindowRepository timewindowRepo;

    @Autowired
    ItemRepository itemsRepo;

    public static final String ORDER_ABSOLUTE = "absolute";
    public static final String ORDER_GROUPED = "grouped";
    public static final String ORDER_ROOMS = "rooms";

    public void save(OrderEntity order) {
        repo.save(order);
    }

    public List<OrderEntity> findAll(String uid) {
        return repo.findAllByUserIdOrderByDateDesc(uid);
    }
    
    public OrderEntity getOne(Long id) {
        return repo.getOne(id);
    }

    public List<OrderEntity> findAllByOpening(Long openingId) {
        return repo.findAllByOpeningId(openingId);
    }

    public void updateOrder(Long id, OrderStatus os) {
        Optional<OrderEntity> order = repo.findById(id);
        if (order.isPresent()) {
            OrderEntity orderEntity = order.get();
            orderEntity.setStatus(os);
            repo.save(orderEntity);
        }
    }

    public Long getCircleIdByOrderId(Long id) {
        Optional<OrderEntity> order = repo.findById(id);
        return order.map(orderEntity -> openingRepo.getOne(orderEntity.getOpeningId()).getCircle().getId()).orElse(null);
    }
    
    public ResponseEntity<String> makeOrder(HttpServletRequest request, Long id, int count, long time, String comment,
            String detailsJson) throws IOException {
        
        UserEntity user = util.getUser(request);
        if (user == null)
            return new ResponseEntity<>("Error 403", HttpStatus.FORBIDDEN);
        
        if (user.getRoom() == null || user.getRoom().isEmpty())
            return new ResponseEntity<>("NO_ROOM_SET", HttpStatus.OK);
        
        OrderEntity order = new OrderEntity(user.getUid(), user.getName(),
                "[" + user.getCardType().name() + "] " + comment,
                detailsJson, user.getRoom());
        order.setIntervalId(time);
        ItemEntity item = itemsRepo.getOne(id);
        
        if (!item.isOrderable() || item.isPersonallyOrderable())
            return new ResponseEntity<>("INTERNAL_ERROR", HttpStatus.OK);
        
        order.setName(item.getName());
        OrderDetails details = CustomComponentType.calculateExtra(detailsJson, order, item);
        order.setOpeningId(openings.findNextOf(item.getCircle().getId()).getId());
        OpeningEntity current = openings.findNextOf(item.getCircle().getId());
        
        if (current == null)
            return new ResponseEntity<>("INTERNAL_ERROR", HttpStatus.OK);
        
        if (current.getOrderStart() > System.currentTimeMillis() || current.getOrderEnd() < System.currentTimeMillis())
            return new ResponseEntity<>("NO_ORDERING", HttpStatus.OK);

        if (count < details.getMinCount())
            count = details.getMinCount();
        else if (count > details.getMaxCount())
            count = details.getMaxCount();

        if (current.getOrderCount() + count > current.getMaxOrder())
            return new ResponseEntity<>("OVERALL_MAX_REACHED", HttpStatus.OK);

        TimeWindowEntity timewindow = timewindowRepo.getOne(time);
        if (!timewindow.getOpening().getId().equals(current.getId()))
            return new ResponseEntity<>("INTERNAL_ERROR", HttpStatus.OK);

        if (timewindow.getNormalItemCount() - count < 0)
            return new ResponseEntity<>("MAX_REACHED", HttpStatus.OK);

        if (order.isExtraTag() && timewindow.getExtraItemCount() - count < 0) {
            return new ResponseEntity<>("MAX_REACHED_EXTRA", HttpStatus.OK);
        }

        if (item.getCategory() != ItemCategory.DEFAULT.getId()) {
            if (item.getCategory() == ItemCategory.ALPHA.getId() && current.getUsedAlpha() >= current.getMaxAlpha()
                    || item.getCategory() == ItemCategory.BETA.getId() && current.getUsedBeta() >= current.getMaxBeta()
                    || item.getCategory() == ItemCategory.GAMMA.getId() && current.getUsedGamma() >= current.getMaxGamma()
                    || item.getCategory() == ItemCategory.DELTA.getId() && current.getUsedDelta() >= current.getMaxDelta()
                    || item.getCategory() == ItemCategory.LAMBDA.getId() && current.getUsedLambda() >= current.getMaxLambda()) {
                return new ResponseEntity<>("CATEGORY_FULL", HttpStatus.OK);
            } else {
                switch (ItemCategory.of(item.getCategory())) {
                    case ALPHA: current.setUsedAlpha(current.getUsedAlpha() + count); break;
                    case BETA: current.setUsedBeta(current.getUsedBeta() + count); break;
                    case GAMMA: current.setUsedGamma(current.getUsedGamma() + count); break;
                    case DELTA: current.setUsedDelta(current.getUsedDelta() + count); break;
                    case LAMBDA: current.setUsedLambda(current.getUsedLambda() + count); break;
                }
            }
        }

        timewindow.setNormalItemCount(timewindow.getNormalItemCount() - count);
        if (order.isExtraTag())
            timewindow.setExtraItemCount(timewindow.getExtraItemCount() - count);
        
        current.setOrderCount(current.getOrderCount() + count);
        
        order.setDate(timewindow.getDate());
        order.setPrice(order.getPrice() * count);
        order.setIntervalMessage(timewindow.getName());
        order.setCancelUntil(current.getOrderEnd());
        order.setCategory(item.getCategory());
        order.setPriority(user.getOrderingPriority());
        order.setCompactName((item.getAlias().isEmpty() ? item.getName() : item.getAlias()) + (count == 1 ? "" : (" x " + count)));
        order.setCount(count);

        timewindowRepo.save(timewindow);
        openings.save(current);
        save(order);
        
        return new ResponseEntity<>("ACK", HttpStatus.OK);
    }

    @Deprecated
    public ResponseEntity<String> cancelOrder(HttpServletRequest request, long id) {
        try {
            OrderEntity order = getOne(id);
            if (!order.getUserId().equals(util.getUser(request).getUid()))
                return new ResponseEntity<>("BAD_REQUEST", HttpStatus.BAD_REQUEST);

            if (order.getStatus() != OrderStatus.ACCEPTED) 
                return new ResponseEntity<>("INVALID_STATUS", HttpStatus.OK);
            
            OpeningEntity opening = openings.getOne(order.getOpeningId());
            if (opening.getOrderEnd() <= System.currentTimeMillis()) 
                return new ResponseEntity<>("ORDER_PERIOD_ENDED", HttpStatus.OK);

            order.setStatus(OrderStatus.CANCELLED);
            int count = order.getCount();

            TimeWindowEntity timeWindow = timewindowRepo.getOne(order.getIntervalId());
            timeWindow.setNormalItemCount(timeWindow.getNormalItemCount() + count);
            if (order.isExtraTag())
                timeWindow.setExtraItemCount(timeWindow.getExtraItemCount() + count);
            
            opening.setOrderCount(opening.getOrderCount() - count);

            if (order.getCategory() != ItemCategory.DEFAULT.getId()) {
                switch (ItemCategory.of(order.getCategory())) {
                    case ALPHA: opening.setUsedAlpha(opening.getUsedAlpha() - count); break;
                    case BETA: opening.setUsedBeta(opening.getUsedBeta() - count); break;
                    case GAMMA: opening.setUsedGamma(opening.getUsedGamma() - count); break;
                    case DELTA: opening.setUsedDelta(opening.getUsedDelta() - count); break;
                    case LAMBDA: opening.setUsedLambda(opening.getUsedLambda() - count); break;
                }
            }

            timewindowRepo.save(timeWindow);
            openings.save(opening);
            save(order);
            
            return new ResponseEntity<>("ACK", HttpStatus.OK);
        } catch (Exception e) {
            return new ResponseEntity<>("BAD_REQUEST", HttpStatus.BAD_REQUEST);
        }
    }

    public List<OrderEntity> findToExport(Long openingId, String orderBy) {
        switch (orderBy) {
            case ORDER_ABSOLUTE:
                return appendArtificialId(repo.findAllByOpeningIdAndStatusNotOrderByPriorityDescDateAsc(
                        openingId, OrderStatus.CANCELLED));

            case ORDER_GROUPED:
                return appendArtificialId(repo.findAllByOpeningIdAndStatusNotOrderByIntervalIdAscPriorityDescDateAsc(
                        openingId, OrderStatus.CANCELLED));

            case ORDER_ROOMS:
                return appendArtificialId(repo.findAllByOpeningIdAndStatusNot(openingId, OrderStatus.CANCELLED)
                        .stream().collect(
                                groupingBy(OrderEntity::getIntervalId, groupingBy(OrderEntity::getRoom)))
                        .entrySet()
                        .stream()
                        .sorted(Comparator.comparing(Map.Entry::getKey))
                        .flatMap(interval -> interval.getValue().values().stream()
                                .sorted((a, b) -> Long.compare(
                                        b.stream().max(Comparator.comparing(OrderEntity::getPriority))
                                                .map(OrderEntity::getPriority).orElse(0),
                                        a.stream().max(Comparator.comparing(OrderEntity::getPriority))
                                                .map(OrderEntity::getPriority).orElse(0)))
                                .flatMap(Collection::stream))
                        .collect(Collectors.toList()));

                default: return appendArtificialId(repo.findAllByOpeningIdAndStatusNotOrderByPriorityDescDateAsc(
                        openingId, OrderStatus.CANCELLED));
        }
    }

    private List<OrderEntity> appendArtificialId(List<OrderEntity> source) {
        int id = 1;
        for (OrderEntity order : source)
            order.setArtificialTransientId(id++);
        return source;
    }

}

File: src/main/kotlin/hu/kirdev/schpincer/service/OrderService.kt
package hu.kirdev.schpincer.service

import hu.kirdev.schpincer.dao.*
import hu.kirdev.schpincer.dto.ManualUserDetails
import hu.kirdev.schpincer.dto.PriceBreakdown
import hu.kirdev.schpincer.model.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.MessageSource
import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Isolation
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional
import java.util.*

private val List<OrderEntity>.highestPriority: Int
    get() = this.maxByOrNull { it.priority }?.priority ?: 1

enum class OrderStrategy(val representation: String) {
    ORDER_ABSOLUTE("absolute"),
    ORDER_GROUPED("grouped"),
    ORDER_ROOMS("rooms");
}

const val RESPONSE_NO_ROOM_SET = "NO_ROOM_SET"
const val RESPONSE_INTERNAL_ERROR = "INTERNAL_ERROR"
const val RESPONSE_NO_ORDERING = "NO_ORDERING"
const val RESPONSE_OVERALL_MAX_REACHED = "OVERALL_MAX_REACHED"
const val RESPONSE_MAX_REACHED = "MAX_REACHED"
const val RESPONSE_MAX_REACHED_EXTRA = "MAX_REACHED_EXTRA"
const val RESPONSE_CATEGORY_FULL = "CATEGORY_FULL"
const val RESPONSE_ACK = "ACK"
const val RESPONSE_MANUAL_FAIL = "MANUAL_FAIL"
const val RESPONSE_BAD_REQUEST = "BAD_REQUEST"
const val RESPONSE_INVALID_STATUS = "INVALID_STATUS"
const val RESPONSE_ORDER_PERIOD_ENDED = "ORDER_PERIOD_ENDED"
const val RESPONSE_TIME_WINDOW_INVALID = "TIME_WINDOW_INVALID"
const val BASE_PRICE = "basePrice"

@Service
open class OrderService {

    @Autowired
    internal lateinit var repo: OrderRepository

    @Autowired
    internal lateinit var openingRepo: OpeningRepository

    @Autowired
    internal lateinit var openings: OpeningService

    @Autowired
    internal lateinit var timeWindowRepo: TimeWindowRepository

    @Autowired
    internal lateinit var itemsRepo: ItemRepository

    @Autowired
    internal lateinit var extrasRepository: ExtrasRepository

    @Autowired
    lateinit var messageSource: MessageSource

    @Transactional
    open fun save(order: OrderEntity) {
        repo.save(order)
    }

    @Transactional(readOnly = true)
    open fun findAll(uid: String): List<OrderEntity> {
        return repo.findAllByUserIdOrderByDateDesc(uid)
    }

    @Transactional(readOnly = true)
    open fun getOne(id: Long): OrderEntity? {
        return repo.getReferenceById(id)
    }

    @Transactional(readOnly = true)
    open fun findAllByOpening(openingId: Long): List<OrderEntity> {
        return repo.findAllByOpeningId(openingId)
    }

    @Transactional
    open fun updateOrder(id: Long, os: OrderStatus) {
        repo.findById(id).ifPresent { orderEntity ->
            orderEntity.status = os
            repo.save(orderEntity)
        }
    }

    @Transactional
    open fun updateOrderComment(id: Long, comment: String) {
        repo.findById(id).ifPresent { orderEntity ->
            orderEntity.comment = orderEntity.comment.substring(0, 5) + comment
            repo.save(orderEntity)
        }
    }

    @Transactional
    open fun updateOrderPrice(id: Long, price: Int) {
        repo.findById(id).ifPresent { orderEntity ->
            orderEntity.price = price
            repo.save(orderEntity)
        }
    }

    @Transactional(readOnly = true)
    open fun getCircleIdByOrderId(id: Long): Long? {
        val order: Optional<OrderEntity> = repo.findById(id)
        return order.map { orderEntity: OrderEntity -> openingRepo.getReferenceById(orderEntity.openingId!!).circle?.id!! }.orElse(null)
    }

    @Transactional(readOnly = false)
    open fun reviewOrder(id: Long, reviewId: Long) {
        val order = getOne(id)!!
        order.reviewId = reviewId
        save(order)
    }

    @Transactional(readOnly = false, isolation = Isolation.SERIALIZABLE, propagation = Propagation.REQUIRES_NEW)
    open fun makeOrder(user: UserEntity, id: Long, itemCount: Int, time: Long, comment: String, detailsJson: String): ResponseEntity<String> {
        val procedure = MakeOrderProcedure(
            user, id, itemCount, time, comment, detailsJson,
            itemsRepo = itemsRepo,
            timeWindowRepo = timeWindowRepo,
            extrasRepository = extrasRepository
        )
        procedure.makeOrder()

        timeWindowRepo.save(procedure.timeWindow)
        openings.save(procedure.current)
        this.save(procedure.order)
        return responseOf(RESPONSE_ACK)
    }

    @Transactional(readOnly = false, isolation = Isolation.SERIALIZABLE, propagation = Propagation.REQUIRES_NEW)
    open fun makeManualOrder(user: UserEntity, id: Long, itemCount: Int, time: Long,
                             comment: String, detailsJson: String, manualUser: ManualUserDetails
    ): ResponseEntity<String> {
        val circleId = itemsRepo.findByIdOrNull(id)?.circle?.id
        if (!((user.permissions.contains("CIRCLE_$circleId")) || user.sysadmin)) {
            return responseOf(RESPONSE_MANUAL_FAIL)
        }

        val procedure = MakeOrderProcedure(
            user, id, itemCount, time, comment, detailsJson,
            itemsRepo = itemsRepo,
            timeWindowRepo = timeWindowRepo,
            extrasRepository = extrasRepository
        )
        procedure.makeOrder(manualUser)

        timeWindowRepo.save(procedure.timeWindow)
        openings.save(procedure.current)
        this.save(procedure.order)
        return responseOf(RESPONSE_ACK)
    }

    @Transactional(readOnly = false, isolation = Isolation.SERIALIZABLE, propagation = Propagation.REQUIRES_NEW)
    open fun cancelOrder(user: UserEntity, id: Long): ResponseEntity<String> {
        val procedure = CancelOrderProcedure(user, id,
                orderRepository = repo,
                openings = openings,
                timeWindowRepo = timeWindowRepo)
        procedure.cancelOrder()

        timeWindowRepo.save(procedure.timeWindow)
        openings.save(procedure.opening)
        this.save(procedure.order)
        return responseOf(RESPONSE_ACK)
    }

    @Transactional(readOnly = false, isolation = Isolation.SERIALIZABLE, propagation = Propagation.REQUIRES_NEW)
    open fun changeOrder(user: UserEntity, id: Long, room: String, comment: String): ResponseEntity<String> {
        val procedure = ChangeOrderProcedure(user, id,
            orderRepository = repo,
            room = room,
            comment = comment)
        procedure.changeOrder()

        this.save(procedure.order)
        return responseOf(RESPONSE_ACK)
    }

    @Transactional(readOnly = true)
    open fun findToExport(openingId: Long, orderBy: String): List<OrderEntity> {
        return when (orderBy) {
            OrderStrategy.ORDER_ABSOLUTE.representation -> appendArtificialId(repo.findAllByOpeningIdAndStatusNotOrderByPriorityDescDateAsc(openingId, OrderStatus.CANCELLED))

            OrderStrategy.ORDER_GROUPED.representation -> appendArtificialId(repo.findAllByOpeningIdAndStatusNotOrderByIntervalIdAscPriorityDescDateAsc(openingId, OrderStatus.CANCELLED))

            OrderStrategy.ORDER_ROOMS.representation -> appendArtificialId(repo.findAllByOpeningIdAndStatusNot(openingId, OrderStatus.CANCELLED)
                    .groupBy { it.intervalId }
                    .entries
                    .map { intervals -> Pair(intervals.key, intervals.value.groupBy { it.room }) }
                    .toList()
                    .sortedBy { it.first }
                    .map { intervals ->
                        intervals.first to intervals.second
                                .toList()
                                .sortedBy { it.second.highestPriority }
                                .flatMap { it.second }
                    }
                    .flatMap { it.second }
            )

            else -> appendArtificialId(repo.findAllByOpeningIdAndStatusNotOrderByPriorityDescDateAsc(openingId, OrderStatus.CANCELLED))
        }
    }

    private fun appendArtificialId(source: List<OrderEntity>): List<OrderEntity> {
        var id = 1
        for (order in source)
            order.artificialTransientId = id++
        return source
    }

    @Transactional(readOnly = false)
    open fun closeAllOrdersInOpening(openingId: Long) {
        val orderEntities = repo.findAllByOpeningId(openingId)
                .filter { it.status == OrderStatus.ACCEPTED
                        || it.status == OrderStatus.INTERPRETED
                        || it.status == OrderStatus.COMPLETED
                        || it.status == OrderStatus.HANDED_OVER }
        orderEntities.forEach { it.status = OrderStatus.SHIPPED }
        repo.saveAll(orderEntities)
    }

    @Transactional(readOnly = false)
    open fun cancelAllOrdersInOpening(openingId: Long) {
        val orderEntities = repo.findAllByOpeningId(openingId)
                .filter { it.status == OrderStatus.ACCEPTED }
        orderEntities.forEach { it.status = OrderStatus.CANCELLED }
        repo.saveAll(orderEntities)
    }

    @Transactional(readOnly = false)
    open fun updateStatus(orderId: Long, status: String) {
        repo.findById(orderId).orElse(null)?.let {
            it.status = OrderStatus[status]
            repo.save(it)
        }
    }

    @Transactional(readOnly = false)
    open fun updateChefComment(orderId: Long, comment: String) {
        repo.findById(orderId).orElse(null)?.let {
            it.chefComment = comment
            repo.save(it)
        }
    }

    @Transactional(isolation = Isolation.SERIALIZABLE, propagation = Propagation.REQUIRES_NEW)
    open fun changeCancelUntilDates(openingId: Long, orderEnd: Long) {
        val affectedOpenings = repo.findAllByOpeningId(openingId)
        affectedOpenings.forEach { it.cancelUntil = orderEnd }
        repo.saveAll(affectedOpenings)
    }


    @Transactional(readOnly = true)
    open fun generatePriceBreakdowns(orders: List<OrderEntity>): List<PriceBreakdown> {

        return orders.map {
            val prices = mutableMapOf<String, Int>()
            prices[BASE_PRICE] = it.price / it.count.coerceAtLeast(1)
            it.orderedItem?.apply {
                for (extra in it.extras.sortedBy { extra -> extra.name }) {
                    prices["${extra.name}- ${extra.displayName}"] = extra.price
                }
                prices[BASE_PRICE] = if (this.discountPrice == 0) this.price else this.discountPrice
            }

            PriceBreakdown(it.id, prices)
        }

    }
}

fun responseOf(body: String, status: HttpStatus = HttpStatus.OK) = ResponseEntity(body, status)
